doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+# Dynamic resource sharing for sources and sinks
+
+## Summary
+
+In Materialize today, each source and sink instance is independent. For example:
+
+  * Each instance of a Kafka source creates its own Kafka consumer.
+  * Each instance of a file source creates its own file handle and filesystem
+    event listener.
+  * Each instance of a `TAIL` sink performs its own consolidation of the input
+    stream.
+
+The result is a system that can correctly backfill data when a source is
+instantiated multiple times, but is inefficient in the steady state. In an ideal
+system, once two instances of a source or sink had caught up to the same point,
+they would share resources. To continue the example from above:
+
+  * Multiple instances of a Kafka source would share the underlying Kafka
+    consumer.
+  * Multiple instances of a file source would share the underlying file handler
+    and filesystem event listener.
+  * Multiple instances of a `TAIL` sink would share the underlying consolidation
+    operator.
+
+The proposal here is to teach Materialize to perform this resource sharing
+automatically and dynamically as the catalog evolves.
+
+## Goals
+
+There are three goals here:
+
+  * To raise awareness of the general problem of steady-state inefficiency.
+  * To build consensus that dynamic resource sharing is the general solution to
```

Comment:
Feel free to disregard if this is overly nitpicky!

I understand that this is intended to be generic and high-level, but I found it very difficult to follow this doc and would've had an easier time if it were more concrete. I'm not sure 'dynamic resource sharing' is a good term, or that stuffing eg. TAIL sinks under the umbrella makes it easier to follow.

dynamic resource sharing strikes of some kind of active resource management/time slicing, a la CPU scheduling. It didn't help my understanding because I didn't know specifically what it meant until reading the entirety of the design doc. It may benefit from a much more concrete term, like 'source reader/ingest consolidation' or similar.

Possibly I'm just looking forward to reading the more concrete docs on kafka/postgres cdc consolidation. 🙂